app/src/main/java/com/arctouch/codechallenge/movies_list/presentation/MoviesListPresenter.java

```diff
@@ -0,0 +1,227 @@
+package com.arctouch.codechallenge.movies_list.presentation;
+
+import android.content.Context;
+import android.util.Log;
+
+import com.arctouch.codechallenge.BasePresenter;
+import com.arctouch.codechallenge.R;
+import com.arctouch.codechallenge.api.TmdbApi;
+import com.arctouch.codechallenge.data.Cache;
+import com.arctouch.codechallenge.di.InjectionTmdbApi;
+import com.arctouch.codechallenge.model.Genre;
+import com.arctouch.codechallenge.model.GenreResponse;
+import com.arctouch.codechallenge.model.Movie;
+import com.arctouch.codechallenge.model.UpcomingMoviesResponse;
+
+import java.io.IOException;
+import java.net.SocketTimeoutException;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Objects;
+
+import io.reactivex.Observer;
+import io.reactivex.Scheduler;
+import io.reactivex.disposables.Disposable;
+import retrofit2.Response;
+
+public class MoviesListPresenter extends BasePresenter<MoviesListContract.View> implements MoviesListContract.Presenter {
+
+    private Context context;
+    private Scheduler scheduler;
+    private Scheduler observer;
+
+    public MoviesListPresenter(Context context, MoviesListContract.View view, Scheduler scheduler, Scheduler observer) {
+        super(view);
+        this.context = context;
+        this.scheduler = scheduler;
+        this.observer = observer;
+    }
+
+    @Override
+    public void getGenres(boolean hideProgress) {
+        view.showProgress();
+
+        InjectionTmdbApi.inject()
+                .genres(TmdbApi.API_KEY, TmdbApi.DEFAULT_LANGUAGE)
+                .subscribeOn(scheduler)
+                .observeOn(observer)
+                .subscribe(new Observer<Response<GenreResponse>>() {
+                    @Override
+                    public void onSubscribe(Disposable d) {}
+
+                    @Override
+                    public void onNext(Response<GenreResponse> genreResponseResponse) {
+                        if(hideProgress)
+                            view.hideProgress();
+
+                        if(genreResponseResponse.isSuccessful()){
+                            if(genreResponseResponse.body() != null)
+                                view.onSuccessGenres(genreResponseResponse.body().genres);
```

Comment:
Why do you need to pass genres to the view?

Reply:
It was implemented this way just for separating the call to obtain Genres and the one to obtain Movies. That’s what I thought first. After the test, I checked that I could have done something using RxJava, like with FlatMap to put the results together and there would be just one final result with the Movies filled by the specific genres. Another easier but not better way would be calling a method like `getMovies()`, which in the Presenter calls the `getGenres()` first and, with its result (`onSuccess`), send it to another method of the Presenter like `getMoviesWithGenres()`, and this method would be the responsible by obtaining the Movies and adding the genres to each respective Movie.